### ISSUE TITLE
Use MPTCP to enhance reliability

### ISSUE THREAD

**Issue by Aperence** [Infomaniak/ios-kDrive#1244]
**Is your feature request related to a problem? Please describe.**
Not related to a problem

**Describe the solution you'd like**
Use Multipath TCP (MPTCP) in handover mode to automatically transition to the cellular interface when the connection deteriorates to improve reliability, or another possibility could be to use the aggregate mode to improve the available bandwidth.

**Additional context**
See https://developer.apple.com/documentation/foundation/urlsessionconfiguration/improving_network_reliability_using_multipath_tcp and https://www.mptcp.dev to learn more about the benefits of using MPTCP

**Comment by adrien-coye** [comment]
Hey @Aperence,

Thanks for the interest in our project. We discussed this with the team.
There could be a couple of places where the automatic transition would be useful, but we then have an issue with users with limited data plans as network calls starts on both interfaces. This is still been discussed as we speak.

Regarding the aggregate mode, a modern iPhone will make good use of many cores and can heat up quite a lot while uploading a lot of files. (We have a system to throttle the upload and download speeds if it gets too hot by the way). 

Using both radios (5G + Wi-Fi) raises two issues, as the 5G modem and transceiver does consume quite some power:

- High energy consumption. This goes against our goals of efficiency and eco conception.
- iPhone heating more, possibly causing throttling depending on the environment, thus decreasing overall transfer performance.

I'll add a comment in the issue if we end up using it somewhere. Closing it in the meantime.

Have a good one.

**Comment by Aperence** [comment]
Hello,

Thank you for your answer, I just wanted to argue with some of your arguments. Feel free to take count of them, or to simply ignore them :D

About the power concern, using MPTCP in handover mode should be in most case not a problem. Indeed, unless needed in case of connection deterioration, the path manager of MPTCP on iOS shouldn't create new paths, which results in an negligible overhead (some additional bytes added per packets) compared to classical TCP. Handover mode should thus even help to reduce consumption as we won't need to restart the whole data transfer if the connection is lost. Note that this is only valid on iOS where mobility may be a concern.

For the aggregate mode, it might be interesting to provide an option to the user with a warning, allowing them to be conscious that enabling it might help them to have faster download, but will consume their data. Transferring in parallel on both interfaces should also allow to reduce the transfer time, thus could potentially reduce the overheat as the radio chip will be used for less time. Perhaps also use this mode only when the app is in the foreground, where the radio consumption will be more negligible compared to the screen consumption, which may result in a reduction of the screen time, and thus a reduction of power consumption ? This should of course be tested (perhaps with usage statistics ?) to know the real impact of using the aggregate mode.

Apple also use it for this particular reason, resulting in a better energy consumption https://www.tessares.net/apples-mptcp-story-so-far/

Anyway, thanks for your reply to this issue and have a nice day !

**Comment by matttbe** [comment]
Hello,

@Aperence: thank you for having shared this, and your reply!

@adrien-coye: thank you for your reply!

(I'm part of the team working on MPTCP in the Linux kernel, and I mainly wanted to react about the power consumption.)

> but we then have an issue with users with limited data plans as network calls starts on both interfaces

From what I understood from Apple's devs, the handover mode is quite conservative in terms of power usage, and limited by WiFi Assist. It should not increase the cellular usage, and it could even reduce it, as @Aperence mentioned, because the connections can switch to WiFi when available, and they don't have to be restarted when WiFi is no longer available (when cellular was going to be used anyway).

Note that WiFi assist ensures the same rules are applied: if the user doesn't want to use the cellular, it will not be used with MPTCP as well.

> Using both radios (5G + Wi-Fi) raises two issues, as the 5G modem and transceiver does consume quite some power:

Power usage is usually high when the signal is not good, where there are losses, and retransmissions are needed: in this case the _handover_ mode helps to reduce the power consumption by avoiding wasting time trying to use a "dying" link at full power. It might be good to do some experimentation there: I know some (private) studies have proven that when the iPhone is used in a mobility use-case, MPTCP helps to reduce the battery usage (but I don't have the details).

So in other words, MPTCP does not automatically mean higher energy consumption :)

**Comment by adrien-coye** [comment]
Hey @Aperence and @matttbe 

Thanks for the feedback 👍 

I do understand that MTCP is more efficient than I anticipated (especially the Wi-Fi assist part). 
Also on the iOS side it would indeed be relatively easy to enable.

If we were to do it, I would actually love to experiment `aggregate mode` with a cheap USB-C power meter, on a first sync of an entire photo library. See how the app behaves.

We did investigate further this issue with the backend team, and they thought about it.
There are still roadblocks on the server side to enable this tech. They have no project to enable it as of today.
This would be the real blocker for this tech right now.

Sorry for the negative answer yet again, but I hope both of you have a nice day.

**Comment by matttbe** [comment]
Hi @adrien-coye,

Thank you for your feedback!

> There are still roadblocks on the server side to enable this tech. They have no project to enable it as of today.
> This would be the real blocker for this tech right now.

Do you think it is possible to know more about what is blocking MPTCP to be deployed on the server side?

I mean: I'm not here to push you to use this technology, you are free to use it or not. It is just that I'm part of the team maintaining MPTCP in the Linux kernel, and that's always good to get some feedback, and find out what is blocking people to use MPTCP. There are maybe stuff that we can improve :)

**Comment by adrien-coye** [comment]
> Do you think it is possible to know more about what is blocking MPTCP to be deployed on the server side?

Yup, I asked validation for this.
It is not natively supported by HAProxy, but there are ways to do it that would require some work for the infrastructure team, it was thus not declared a priority.

I just found out [this PR](https://github.com/haproxy/haproxy/issues/1028) that you are probably aware of.

**Comment by matttbe** [comment]
> Hey @matttbe
> 
> > Do you think it is possible to know more about what is blocking MPTCP to be deployed on the server side?
> 
> Yup, I asked validation for this. It is not natively supported by HAProxy, but there are ways to do it that would require some work for the infrastructure team, it was thus not declared a priority.

Thank you for your reply!

Yes, indeed, HAProxy doesn't support MPTCP natively. (Note that it is possible to force application to use MPTCP instead of TCP, see [here](https://www.mptcp.dev/setup.html#force-applications-to-use-mptcp) for more details about that.)

> I just found out [this PR](https://github.com/haproxy/haproxy/issues/1028) that you are probably aware of.

Funny, @Aperence has recently taken over a work I started a few months ago, and sent patches adding native MPTCP support in HAProxy a few hours ago: https://www.mail-archive.com/haproxy@formilux.org/msg45253.html
